### PR TITLE
fix(docker): use a GET request to the real resource auth. (#14744)

### LINKED ISSUE #14708: ACR Datasource unauthorized while fetching tags list.
### How are you running Renovate?

Self-hosted

### If you're self-hosting Renovate, tell us what version of Renovate you run.

32.6.0 

### Please select which platform you are using if self-hosting.

GitLab self-hosted

### If you're self-hosting Renovate, tell us what version of the platform you run.

14.8.2

### Was this something which used to work for you, and then stopped?

I never saw this working

### Describe the bug

Using ACR. Fetching new docker image tags is failing.
```
DEBUG: Datasource unauthorized (repository=MY/repository)
       "datasource": "docker",
       "packageName": "redacted.azurecr.io/upstream/rust",
       "url": "https://redacted.azurecr.io/v2/upstream/rust/tags/list?n=10000"
       
```
I tried to reproduced the call and authentication performed by renovate.
I'm facing the same authentication issue.

```
{
    "errors": [
        {
            "code": "UNAUTHORIZED",
            "message": "authentication required, visit https://aka.ms/acr/authorization for more information.",
            "detail": [
                {
                    "Type": "repository",
                    "Name": "upstream/rust",
                    "Action": "metadata_read"
                }
            ]
        }
    ]
}
```
Renovate is requesting a token with the following scope. `repository:upstream/rust:pull`.

Scope should be either `repository:upstream/rust:pull,metadata_read` or `repository:upstream/rust:*`

### Relevant debug logs

<details><summary>Logs</summary>

```
Copy/paste any log here, between the starting and ending backticks
```

</details>

### Have you created a minimal reproduction repository?

No reproduction repository

### PR DESCRIPTION
**Pull request by vlagorce**
## Changes

Some docker registries needs a specific scope when listing available tags. 
This change aims to pass the registry url to the "auth" method in order to retrieve the right scope.

## Context

Closes #14708

## Documentation (please check one with an [x])

- [ ] I have updated the documentation, or
- [x] No documentation update is required

## How I've tested my work (please tick one)

I have verified these changes via:

- [ ] Code inspection only, or
- [x] Newly added/modified unit tests, or
- [ ] No unit tests but ran on a real repository, or
- [x] Both unit tests + ran on a real repository

### CODE REVIEW

**Comment by viceice** [comment]
please add all relevant issues / PRs to context section

File: lib/modules/datasource/docker/index.ts [hunk 1]
```diff
@@ -53,25 +53,25 @@ function isDockerHost(host: string): boolean {
...
   try {
-    const apiCheckUrl = `${registryHost}/v2/`;
     const apiCheckResponse = await http.get(apiCheckUrl, {
       throwHttpErrors: false,
       noAuth: true,
     });
 
     if (apiCheckResponse.statusCode === 200) {
-      logger.debug({ registryHost }, 'No registry auth required');
+      logger.debug({ apiCheckUrl }, 'No registry auth required');
```
Comment (JamieMagee) [inline lib/modules/datasource/docker/index.ts:66]: I think it's still useful to have the `registryHost` here to help with debugging.

**Comment by JamieMagee** [comment]
Which container registries did you test against?

File: lib/modules/datasource/docker/index.ts (hunk 1 above)
Comment (vlagorce) [reply lib/modules/datasource/docker/index.ts:66]: apiCheckUrl will always contains the registryHost.

File: lib/modules/datasource/docker/index.spec.ts [hunk 2]
```diff
@@ -580,7 +615,7 @@ describe('modules/datasource/docker/index', () => {
       const tags = ['1.0.0'];
       httpMock
         .scope('https://registry.company.com/v2')
-        .get('/')
+        .get('/node/tags/list?n=10000')
```
Comment (viceice) [inline lib/modules/datasource/docker/index.spec.ts:618]: it seems we do the get request twice when no auch is required? this shouldn't happen and the request should be cached in memory.

🤔 @rarkins is request caching somehow disabled on tests?

Comment (viceice) [reply lib/modules/datasource/docker/index.spec.ts:618]: yes, so i thinking the cache should have catched the second call below 🤔

Comment (vlagorce) [reply lib/modules/datasource/docker/index.spec.ts:618]: This behavior was present before this MR. 
It is emphasize by using a specific path when getting the Auth header in order to fetch the tag list.

Before the MR
`.get('/')` is the call to get the header `getAuthHeaders`
`.get('/node/tags/list?n=10000')` is the call to get the tags  `getDockerApiTags`

The MR
`.get('/node/tags/list?n=10000')` is the call to get the header  `getAuthHeaders`
`.get('/node/tags/list?n=10000')` is the call to get the tags `getDockerApiTags`

Regarding the cache.
If I'm not missing some code, caching strategy is define at the method level not the remote api call level.

Comment (vlagorce) [reply lib/modules/datasource/docker/index.spec.ts:618]: Anything to fix change ?

Comment (viceice) [reply lib/modules/datasource/docker/index.spec.ts:618]: Ok, i see the issue, memcache isn't enabled. we should add a test with enabled cache

https://github.com/renovatebot/renovate/blob/637585c5fd9f7d3239318ad9e9e3ec288ec3c6e1/lib/util/http/index.ts#L149-L155

Comment (vlagorce) [reply lib/modules/datasource/docker/index.spec.ts:618]: hi @viceice,
 
Should I add memcache testing in this PR ? 
I feel  like testing this should be a full task in order to properly cover all accesses.

Comment (viceice) [reply lib/modules/datasource/docker/index.spec.ts:618]: We can defer cache test to a later pr.

**Review by viceice — APPROVED** [review]

**Pull request merged**